### PR TITLE
Complete rewrite of bsd IClient

### PR DESCRIPTION
**Pull request by Thog**
This should provide more accurate behaviours than before.
This implementation has been tested with ftpd and libtransistor bsd tests.

This implementation lacks OOB support.

### CODE REVIEW

**Comment by Thog** [comment]
Tested and working like a charm on Mnecraft Bedrock Edition

File: Ryujinx.HLE/HOS/Services/Bsd/BsdSocket.cs [hunk 1]
```diff
@@ -9,10 +9,6 @@ class BsdSocket
         public int Type;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/BsdSocket.cs]: `using System.Net;` at the beginning of the file is unneeded, could you remove it please!

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 2]
```diff
@@ -14,32 +13,149 @@ class IClient : IpcService
...
+        private long WriteBsdResult(ServiceCtx Context, int Result, BsdError ErrorCode = 0)
+        {
+            if (ErrorCode != BsdError.Success)
+            {
+                Result = -1;
+            }
+
+            Context.ResponseData.Write(Result);
+            Context.ResponseData.Write((int)ErrorCode);
+            return 0;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Add an empty line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 3]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
-            int SocketId    = Context.RequestData.ReadInt32();
-            int SocketFlags = Context.RequestData.ReadInt32();
+            int         SocketFd    = Context.RequestData.ReadInt32();
+            SocketFlags SocketFlags = (SocketFlags)Context.RequestData.ReadInt32();
 
-            byte[] SentBuffer = Context.Memory.ReadBytes(Context.Request.SendBuff[0].Position,
-                                                         Context.Request.SendBuff[0].Size);
+            (long ReceivePosition,     long ReceiveLength)   = Context.Request.GetBufferType0x22();
+            (long SockAddrInPosition,  long SockAddrInSize)  = Context.Request.GetBufferType0x21();
+            (long SockAddrOutPosition, long SockAddrOutSize) = Context.Request.GetBufferType0x22(1);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Remove extra empty line.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 4]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
+            int       Result = -1;
 
-            if (!Sockets[SocketId].Handle.Connected)
+            if (Socket != null)
             {
+                if (SocketFlags != SocketFlags.None && (SocketFlags & SocketFlags.OutOfBand) == 0
+                    && (SocketFlags & SocketFlags.Peek) == 0)
+                {
+                    Context.Device.Log.PrintWarning(Logging.LogClass.ServiceBsd, $"Unsupported Recv flags: {SocketFlags}");
+                    return WriteBsdResult(Context, -1, BsdError.OperationNotSupported);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Add an empty line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 5]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
+
+            if (Socket != null)
             {
-                Context.ResponseData.Write(-1);
-                Context.ResponseData.Write(Ex.ErrorCode - 10000);
+                if (SocketFlags != SocketFlags.None && SocketFlags != SocketFlags.OutOfBand
+                    && SocketFlags != SocketFlags.Peek && SocketFlags != SocketFlags.DontRoute)
+                {
+                    Context.Device.Log.PrintWarning(Logging.LogClass.ServiceBsd, $"Unsupported Send flags: {SocketFlags}");
+                    return WriteBsdResult(Context, -1, BsdError.OperationNotSupported);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Add an empty line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 6]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
+            int       Result = -1;
 
-            Task TimeOut = Task.Factory.StartNew(() =>
+            if (Socket != null)
             {
+                if (SocketFlags != SocketFlags.None && SocketFlags != SocketFlags.OutOfBand
+                    && SocketFlags != SocketFlags.Peek && SocketFlags != SocketFlags.DontRoute)
+                {
+                    Context.Device.Log.PrintWarning(Logging.LogClass.ServiceBsd, $"Unsupported Send flags: {SocketFlags}");
+                    return WriteBsdResult(Context, -1, BsdError.OperationNotSupported);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Add an empty line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 7]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
-                    Writer.Write((byte)NewBsdSocket.Handle.AddressFamily);
+                if (NewSocket == null && Errno == BsdError.Success)
+                {
+                    Errno = BsdError.Again;
+                }
+                else if (Errno == BsdError.Success)
+                {
+                    BsdSocket NewBsdSocket = new BsdSocket
+                    {
+                        Family     = (int)NewSocket.AddressFamily,
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Remove extra spaces before assign.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 8]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
+                        // FIXME: OOB not implemented.
+                        Context.Memory.WriteInt32(BufferPosition, 0);
+                        break;
+
+                    default:
+                        Errno = BsdError.OperationNotSupported;
+
+                        Context.Device.Log.PrintWarning(Logging.LogClass.ServiceBsd, $"Unsupported Ioctl Cmd: {Cmd}");
+                        break;
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Remove empty line.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 9]
```diff
@@ -54,447 +170,910 @@ public long Initialize(ServiceCtx Context)
...
+                    case SocketOptionName.ReceiveBuffer:
+                    case SocketOptionName.ReceiveTimeout:
+                    case SocketOptionName.SendBuffer:
+                    case SocketOptionName.SendTimeout:
+                    case SocketOptionName.Type:
+                    case SocketOptionName.Linger:
+                        Socket.Handle.GetSocketOption(SocketOptionLevel.Socket, OptionName, OptionValue);
+                        Context.Memory.WriteBytes(OptionValuePosition, OptionValue);
+
+                        return BsdError.Success;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Add an empty line after.

File: Ryujinx.HLE/HOS/Services/Bsd/PollEvent.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Bsd/PollEvent.cs]: Remove Unneeded `using`.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 11]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+                }
+
+                if (ReadEvents.Contains(Socket))
+                {
+                    if ((Event.InputEvents & PollEvent.EventTypeMask.Input) != 0)
+                    {
+                        OutputEvents |= PollEvent.EventTypeMask.Input;
+                    }
+                }
+                if (WriteEvents.Contains(Socket))
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: Missing empty new line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 12]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+                {
+                    OutputEvents |= PollEvent.EventTypeMask.Output;
+                }
+                Context.Memory.WriteInt16(BufferPosition + i * 8 + 6, (short)OutputEvents);
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: Missing empty new line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 13]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
-            byte[] AddressBuffer = Context.Memory.ReadBytes(AddressPosition, AddressSize);
+            BsdError  Errno  = BsdError.BadFileDescriptor;
+            BsdSocket Socket = RetrieveSocket(SocketFd);
 
-            try
+            if (Socket != null)
             {
-                ParseAddrBuffer(SocketId, AddressBuffer);
+                Errno = BsdError.Success;
+                try
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: Missing empty new line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 14]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+                        (long BufferPosition, long BufferSize) = Context.Request.GetBufferType0x22();
+
+                        // FIXME: OOB not implemented.
+                        Context.Memory.WriteInt32(BufferPosition, 0);
+                        break;
+
+                    default:
+                        Errno = BsdError.OperationNotSupported;
+
+                        Context.Device.Log.PrintWarning(Logging.LogClass.ServiceBsd, $"Unsupported Ioctl Cmd: {Cmd}");
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Hmm, we usually throw NotImplementedException for unimplemented stuff, unless its something very common that would block further testing on games.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 15]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+            int       Result = 0;
+            BsdError  Errno  = BsdError.BadFileDescriptor;
+            BsdSocket Socket = RetrieveSocket(SocketFd);
 
-            byte[] SocketOptionValue = Context.Memory.ReadBytes(Context.Request.PtrBuff[0].Position,
-                                                                Context.Request.PtrBuff[0].Size);
+            if (Socket != null)
+            {
+                Errno = BsdError.Success;
+                if (Cmd == 0x3)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: Missing empty new line before.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 16]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+            ulong Reserved = Context.RequestData.ReadUInt64();
+
+            BsdError Errno     = BsdError.OperationNotPermitted;
+            int      NewSockFd = -1;
+
+            if (IsPrivileged)
+            {
+                Errno = BsdError.BadFileDescriptor;
+                BsdSocket OldSocket = RetrieveSocket(SocketFd);
+                if (OldSocket != null)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: Missing empty new line before.

File: Ryujinx.HLE/HOS/Services/Bsd/BsdError.cs [hunk 17]
```diff
@@ -1,8 +1,19 @@
 namespace Ryujinx.HLE.HOS.Services.Bsd
 {
+    // FIXME(thog): this assumption might be not accurate at all.
     //bsd_errno == (SocketException.ErrorCode - 10000)
     enum BsdError
```
Comment (jduncanator) [inline Ryujinx.HLE/HOS/Services/Bsd/BsdError.cs]: This enum, and the above assumption are incorrect.

A number of these values (when you add 10000 to them) don't even map to .NET SocketError values:

`Success`
`OperationNotPermitted`
`NotFound`
`BadFileDescriptor`
`Again`
`OperationNotSupported`

The following values map to (incorrect?) different values in .NET:

`ProtocolNotAvailaible` is currently mapped to `SocketError.VersionNotSupported`
`ProtocolNotSupported` is currently mapped to `SocketError.NotInitialized`

On top of this, there are a number of .NET errors that could probably map to a single BsdError, so I think a more advanced mapping scheme is required here.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 18]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+                PollEvent.EventTypeMask OutputEvents = 0;
+
+                Socket Socket = Event.Socket.Handle;
+
+                if (ErrorEvents.Contains(Socket))
+                {
+                    OutputEvents |= PollEvent.EventTypeMask.Error;
+                    if (!Socket.Connected || !Socket.IsBound)
+                    {
+                        OutputEvents |= PollEvent.EventTypeMask.Disconnected;
```
Comment (jduncanator) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: Do you need to unset the Error event mask here?

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 19]
```diff
@@ -54,447 +171,912 @@ public long Initialize(ServiceCtx Context)
...
+                try
+                {
+                    Result = Socket.Handle.Receive(ReceivedBuffer, SocketFlags);
+                    Errno  = SetResultErrno(Socket.Handle, Result);
+
+                    Context.Memory.WriteBytes(ReceivePosition, ReceivedBuffer);
+                }
+                catch (SocketException Exception)
+                {
+                    Errno = (BsdError)Exception.ErrorCode - 10000;
```
Comment (jduncanator) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: See previous comments re. error codes.

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 20]
```diff
@@ -54,447 +273,919 @@ public long Initialize(ServiceCtx Context)
...
+                Context.Memory.WriteInt16(BufferPosition + i * 8 + 4, (short)Event.InputEvents);
+
+                PollEvent.EventTypeMask OutputEvents = 0;
+
+                Socket Socket = Event.Socket.Handle;
+
+                if (ErrorEvents.Contains(Socket))
+                {
+                    OutputEvents |= PollEvent.EventTypeMask.Error;
+                    if (!Socket.Connected || !Socket.IsBound)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: nit: New line before.

**Comment by gdkchan** [comment]
Build is failing, needs a rebase due to the recent logger changes.

**Comment by Thog** [comment]
Rebased

File: Ryujinx.HLE/HOS/Services/Bsd/IClient.cs [hunk 21]
```diff
@@ -54,447 +274,920 @@ public long Initialize(ServiceCtx Context)
...
+        public long Fcntl(ServiceCtx Context)
         {
-            int SocketId = Context.RequestData.ReadInt32();
+            int SocketFd = Context.RequestData.ReadInt32();
+            int Cmd      = Context.RequestData.ReadInt32();
+            int Arg      = Context.RequestData.ReadInt32();
 
-            SocketOptionLevel SocketLevel     = (SocketOptionLevel)Context.RequestData.ReadInt32();
-            SocketOptionName SocketOptionName = (SocketOptionName)Context.RequestData.ReadInt32();
+            int       Result = 0;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Bsd/IClient.cs]: This is misaligned.

**Comment by trevor403** [comment]
This implementation seems to only work in libtransisor and not in libnx. I will do some more digging and open an Issue.

**Comment by trevor403** [comment]
This test works in libnx https://gist.github.com/kbhomes/5165d57df184e6dd1c4f7ad39584990f
Note the Makefile is out of date.

**Pull request merged**